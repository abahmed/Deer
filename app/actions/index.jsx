// Local libraries
import { ACTION_TYPES } from './types';

/*
*  All notes related actions including payloads, if any
*/

// fetchAllNotes - Retrieve notes from the store

export function fetchAllNotes() {
  return {
    type: ACTION_TYPES.FETCH_ALL_NOTES,
  };
}

// addNote - Add/Save a new note to the store
export function addNote(payload) {
  return {
    type: ACTION_TYPES.ADD_NOTE,
    payload: payload,
  };
}

// const ACTIONS = {
//   addNote: addNote,
//   fetchAllNotes: fetchAllNotes
// }

// export default ACTIONS;

