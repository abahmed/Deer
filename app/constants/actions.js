export const ACTIONS = {
  GET_NEXT_LANG: 'SET_NEXT_LANG',
  TOGGLE_FADE: 'TOGGLE_FADE',
  UPDATE_EDITOR_STATE: 'UPDATE_EDITOR_STATE',
  FETCH_ALL_NOTES: 'FETCH_ALL_NOTES',
  UPDATE_NOTE_LIST: 'UPDATE_NOTE_LIST',
  SET_ACTIVE_NOTE: 'SET_ACTIVE_NOTE',
  UPDATE_NEW_NOTE_STATE: 'UPDATE_NEW_NOTE_STATE',
  UPDATE_SAVE_NOTE_STATE: 'UPDATE_SAVE_NOTE_STATE',
  ADD_NOTE: 'ADD_NOTE'
}
