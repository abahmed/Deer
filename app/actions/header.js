import { createAction } from 'redux-actions'
import { ACTIONS } from '../constants/actions'

// Used for updating state of adding new note button (e.g. enabled or disabled).
export const setNewNoteDisabled = createAction(ACTIONS.UPDATE_NEW_NOTE_STATE)

// Used for updating state of save button (e.g. enabled or disabled).
export const setSaveDisabled = createAction(ACTIONS.UPDATE_SAVE_NOTE_STATE)

// Used for updating state of deleteing note button (e.g. enabled or disabled).
export const setDeleteDisabled = createAction(ACTIONS.UPDATE_DELETE_NOTE_STATE)
