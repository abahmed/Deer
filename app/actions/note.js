import { createAction } from 'redux-actions'
import { ACTIONS } from '../constants/actions'
import { fetchNotes } from './../db'

// Used for adding a new note.
export const addNewNote = createAction(ACTIONS.ADD_NOTE)

// Used for updating noteList with fetched notes from database.
export const updateNoteList = createAction(ACTIONS.UPDATE_NOTE_LIST)

// Used for setting the id of selected note.
export const setActiveNote = createAction(ACTIONS.SET_ACTIVE_NOTE)

// Used for updating title of active note.
export const updateNoteTitle = createAction(ACTIONS.UPDATE_NOTE_TITLE)

// Used for fetching all notes from database.
export const fetchAllNotes = () => (dispatch, getState) => {
  fetchNotes().then((result) => {
    // Update only if there are notes.
    if (result.total_rows > 0) {
      dispatch(updateNoteList(result.rows))
    }
  }).catch((err) => {
    console.error(err)
  })
}
