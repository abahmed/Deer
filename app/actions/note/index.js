import { createAction } from 'redux-actions'
import { ACTIONS } from '../../constants/actions'
import { addNote, fetchNotes, removeNote } from '../../utils/db'
import logger from 'electron-log'

/** Used for adding a new note. */
const addNewNote = createAction(ACTIONS.ADD_NOTE)

/** Used for updating notes with fetched notes from database. */
const updateNotesList = createAction(ACTIONS.NOTES_FETCH_SUCCESS)

/** Used for setting the index of selected note. */
const setSelectedNoteID = createAction(ACTIONS.SET_SELECTED_NOTE_ID)

/** Used for updating title, content and modified values for the selected note. */
const editSelectedNote = createAction(ACTIONS.EDIT_SELECTED_NOTE)

/** Used for updating rev of the selected note. */
const updateSelectedNoteRev = createAction(ACTIONS.UPDATE_SELECTED_NOTE_REV)

/** Used for deleting a note from noteList. */
const deleteSelectedNote = createAction(ACTIONS.DELETE_SELECTED_NOTE)

/** Async method, Used for fetching all notes from database. */
const fetchAllNotes = () => (dispatch) => {
  fetchNotes().then((result) => {
    // Update only if there are notes.
    if (result.total_rows > 0) {
      dispatch(updateNotesList(result.rows))
    }
  }).catch((err) => {
    logger.error('Unable to fetch notes: ' + JSON.stringify(err))
  })
}

/**
 * Async method, used for saving note to database.
 */
const createNote = () => (dispatch) => {
  const modified = Date.now()
  addNote(undefined, undefined, undefined, modified).then((result) => {
    if (result.ok) {
      dispatch(addNewNote({
        id: result.id,
        rev: result.rev,
        title: '',
        content: '',
        modified: modified
      }))
      dispatch(setSelectedNoteID(result.id))
    }
  }).catch((err) => {
    logger.error('Unable to add new note: ' + JSON.stringify(err))
  })
}

/**
 * Async method, used for updating note in database.
 */
const saveSelectedNote = () => (dispatch, getState) => {
  const state = getState().noteReducer
  const selectedNoteID = state.get('selectedNoteID')
  const selectedNote = state.get('notes')[selectedNoteID]
  addNote(
    selectedNoteID,
    selectedNote.title,
    selectedNote.content,
    selectedNote.modified,
    selectedNote.rev
  ).then((result) => {
    if (result.ok) {
      // Update note rev to avoid conflicts while the next time for saving
      // this note.
      dispatch(updateSelectedNoteRev(result.rev))
    }
  }).catch((err) => {
    logger.error('Unable to update note: ' + JSON.stringify(err))
  })
}


/** Async method, used for removing active note from database. */
const removeSelectedNote = () => (dispatch, getState) => {
  const state = getState().noteReducer
  const selectedNoteID = state.get('selectedNoteID')
  const selectedNote = state.get('notes')[selectedNoteID]
  removeNote(selectedNoteID, selectedNote.rev).then((result) => {
    dispatch(deleteSelectedNote())
  }).catch((err) => {
    logger.error('Unable to remove note ' + JSON.stringify(err))
  })
}

export {
  updateNotesList,
  addNewNote,
  setSelectedNoteID,
  editSelectedNote,
  updateSelectedNoteRev,
  fetchAllNotes,
  createNote,
  saveSelectedNote,
  deleteSelectedNote,
  removeSelectedNote
}
