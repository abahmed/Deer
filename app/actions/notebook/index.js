import { createAction } from 'redux-actions'
import { ACTIONS } from '../../constants/actions'
import { fetchNoteBooks, addNoteBook, removeNoteBook } from '../../utils/db'
import logger from 'electron-log'

/** Used for adding a new notebook. */
const addNewNoteBook = createAction(ACTIONS.ADD_NOTE_BOOK)

/** Used for updating notebooks with fetched notebooks from database. */
const updateNoteBooksList = createAction(ACTIONS.NOTE_BOOKS_FETCH_SUCCESS)

/** Used for setting the index of active notebook. */
const setActiveNoteBookID = createAction(ACTIONS.SET_ACTIVE_NOTE_BOOK_ID)

/** Used for setting note book is opened to true . */
const openNoteBook = createAction(ACTIONS.OPEN_NOTE_BOOK)

/** Used for setting note book is opened to false . */
const closeNoteBook = createAction(ACTIONS.CLOSE_NOTE_BOOK)

/** Used for setting the index of selected notebook. */
const setSelectedNoteBookID = createAction(ACTIONS.SET_SELECTED_NOTE_BOOK_ID)

/** Used for deleting a notebook from noteBookList. */
const deleteSelectedNoteBook = createAction(ACTIONS.DELETE_SELECTED_NOTE_BOOK)

/** Used for creating a list of the active notebook's noteIDs. */
const createNoteList = createAction(ACTIONS.CREATE_NOTE_LIST)

/** Used for deleting a list of the active notebook's noteIDs. */
const deleteNoteList = createAction(ACTIONS.DELETE_NOTE_LIST)

/** Used for adding note to noteIDs. */
const addToNoteIDs = createAction(ACTIONS.ADD_NOTE_TO_NOTE_IDS)

/** Used for removing notes from noteIDs. */
const removeFromNoteIDs = createAction(ACTIONS.REMOVE_NOTE_FROM_NOTE_IDS)

/** Used for adding note to the active notebook. */
const addToActiveNoteBook = createAction(ACTIONS.ADD_NOTE_TO_ACTIVE_NOTE_BOOK)

/** Used for removing notes from the active notebook. */
const removeFromActiveNoteBook = createAction(ACTIONS.REMOVE_NOTE_FROM_ACTIVE_NOTE_BOOK)

/** Used for updating rev of active notebook. */
const updateActiveNoteBookRev = createAction(ACTIONS.UPDATE_ACTIVE_NOTE_BOOK_REV)

/** Used for updating the notebook */
const updateNoteBook = createAction(ACTIONS.UPDATE_NOTE_BOOK)

/** Async method, Used for fetching all notebooks from database. */
const fetchAllNoteBooks = () => dispatch => {
  fetchNoteBooks()
    .then(result => {
      // Update only if there are notebooks.
      if (result.total_rows > 0) {
        dispatch(updateNoteBooksList(result.rows))
      }
    })
    .catch(err => {
      logger.error('Unable to fetch notebooks: ' + JSON.stringify(err))
    })
}

/**
 * Async method, used for saving notebook to database.
 */
const createNoteBook = name => dispatch => {
  const modified = Date.now()
  addNoteBook(undefined, name, undefined, modified)
    .then(result => {
      if (result.ok) {
        dispatch(
          addNewNoteBook({
            id: result.id,
            rev: result.rev,
            name: name,
            noteIDs: [],
            modified: modified
          })
        )
      }
    })
    .catch(err => {
      logger.error('Unable to add new notebook: ' + JSON.stringify(err))
    })
}

/**
 * Async method, used for updating notebook in database.
 */
const saveActiveNoteBook = () => (dispatch, getState) => {
  const state = getState().noteBookReducer
  const activeNoteBookID = state.get('activeNoteBookID')
  const activeNoteBook = state.get('noteBooks')[activeNoteBookID]
  addNoteBook(
    activeNoteBookID,
    activeNoteBook.name,
    activeNoteBook.noteIDs,
    activeNoteBook.modified,
    activeNoteBook.rev
  )
    .then(result => {
      if (result.ok) {
        // Update notebook rev to avoid conflicts while the next time for saving
        // this notebook.
        dispatch(updateActiveNoteBookRev(result.rev))
      }
    })
    .catch(err => {
      logger.error('Unable to update notebook: ' + JSON.stringify(err))
    })
}

/** Async method, used for removing active notebook from database. */
const removeSelectedNoteBook = () => (dispatch, getState) => {
  const state = getState().noteBookReducer
  const selectedNoteBookID = state.get('selectedNoteBookID')
  const selectedNoteBook = state.get('noteBooks')[selectedNoteBookID]
  removeNoteBook(selectedNoteBookID, selectedNoteBook.rev)
    .then(result => {
      dispatch(deleteSelectedNoteBook())
    })
    .catch(err => {
      logger.error('Unable to remove notebook ' + JSON.stringify(err))
    })
}
/** Async method, used for removing the destroyed note from notebook database. */
const removeNoteFromAllNoteBooks = (noteID) => (dispatch, getState) => {
  const state = getState().noteBookReducer
  const noteBooks = state.get('noteBooks')
  Object.keys(noteBooks).map( key => {
    const notebook = noteBooks[key]
    const noteIDs = notebook.noteIDs.filter( id => id !== noteID)
    if(noteIDs.length !== notebook.noteIDs.length){
      const modified = Date.now()
      addNoteBook(
        key,
        notebook.name,
        noteIDs,
        modified,
        notebook.rev
      ).then(result => {
          if (result.ok) {
            dispatch(updateNoteBook({id:key, notebook: {...notebook,noteIDs,modified,rev:result.rev}}))
          }
        })
        .catch(err => {
          logger.error('Unable to update notebook: ' + JSON.stringify(err))
        })
    }
  })
}

export {
  updateNoteBooksList,
  addNewNoteBook,
  fetchAllNoteBooks,
  createNoteBook,
  deleteSelectedNoteBook,
  removeSelectedNoteBook,
  setActiveNoteBookID,
  setSelectedNoteBookID,
  addToNoteIDs,
  removeFromNoteIDs,
  addToActiveNoteBook,
  removeFromActiveNoteBook,
  updateActiveNoteBookRev,
  openNoteBook,
  closeNoteBook,
  createNoteList,
  deleteNoteList,
  saveActiveNoteBook,
  removeNoteFromAllNoteBooks
}
