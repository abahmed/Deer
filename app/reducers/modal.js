import { ACTIONS } from '../constants/actions'

const INITIAL_STATE = {
  showSaveModal: false
}

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case ACTIONS.TOGGLE_SAVE_MODAL:
      return {
        ...state,
        showSaveModal: !state.showSaveModal
      }
    default:
      return state
  }
}
