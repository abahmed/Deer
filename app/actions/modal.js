import { createAction } from 'redux-actions'
import { ACTIONS } from '../constants/actions'
import { NOTE_STATUS } from '../constants/noteStatus'
import { setNoteStatus } from './note'

// Used for updating visibility of save modal.
export const toggleSaveModal = createAction(ACTIONS.TOGGLE_SAVE_MODAL)

// Async method, Used for updating note status with two optional parameters
// - withTimeOut: dispatch action after timeout or not and it's default value
//                is true.
// - status: the new status of the note and it's default value is NO_OPERATION.
export const updateNoteStatus =
  (withTimeOut = true, status = NOTE_STATUS.NO_OPERATION) =>
    (dispatch, getState) => {
      setTimeout(() => {
        dispatch(setNoteStatus(status))
      }, withTimeOut ? 1000 : 0)
    }
