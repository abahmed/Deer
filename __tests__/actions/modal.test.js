import configureStore from 'redux-mock-store'
import { ACTIONS } from '../../app/constants/actions'
import * as modalActions from '../../app/actions/modal'
jest.mock('../../__mocks__/pouchdb-browser')

const mockStore = configureStore()
const store = mockStore()

global.fetch = () => {}

describe('modal_actions', () => {
  beforeEach(() => {
    store.clearActions()
  })

  describe('toggleYesNoModal', () => {
    test('Dispatches the correct action and payload', () => {
      store.dispatch(modalActions.toggleYesNoModal(ACTIONS.SAVE_NOTE))
      expect(store.getActions()).toMatchSnapshot()
    })
  })
})
