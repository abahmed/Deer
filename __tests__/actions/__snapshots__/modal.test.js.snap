// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal_actions toggleYesNoModal Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": "SAVE_NOTE",
    "type": "TOGGLE_SAVE_MODAL",
  },
]
`;
